### PR TITLE
Add App with TTA

### PR DESCRIPTION
**Pull request by diazandr3s**
Signed-off-by: Andres <diazandr3s@gmail.com>

- Create App with Test Time Augmentation (TTA)

### CODE REVIEW

**Comment by diazandr3s** [comment]
@SachidanandAlle just completed the first set of tests for TTA. it is working as planned. Please review it and let me know if it is ready to submit

File: monailabel/utils/scoring/tta_scoring.py [hunk 1]
```diff
@@ -10,45 +10,155 @@
 # limitations under the License.
 
 import logging
+import time
+from functools import partial
 
+import matplotlib.pyplot as plt
```
Comment (SachidanandAlle) [inline monailabel/utils/scoring/tta_scoring.py:16]: lets not have this dependency for production.. i guess it's not required

File: monailabel/utils/scoring/tta_scoring.py [hunk 2]
```diff
@@ -10,45 +10,155 @@
...
+        post_pred = transforms(preds)
+        return post_pred
+
+    def get_2d_im(self, im, channel, z_slice):
+        im = im[..., z_slice]
+        if channel is not None:
+            im = im[channel][None]
+        return im
+
+    def imshows(self, ims):
```
Comment (SachidanandAlle) [inline monailabel/utils/scoring/tta_scoring.py:90]: we can remove this test code...

File: monailabel/utils/scoring/tta_scoring.py [hunk 3]
```diff
@@ -10,45 +10,155 @@
...
+
+            logger.info("Time taken for " + str(self.num_samples) + " augmented samples: " + str(latency_tta))
+
+            # Add vvc in datastore
+            info = {"vvc_tta": vvc_tta}
             logger.info(f"{image_id} => {info}")
             datastore.update_image_info(image_id, info)
             result[image_id] = info
+
+            if self.plot:
```
Comment (SachidanandAlle) [inline monailabel/utils/scoring/tta_scoring.py:147]: same thing..

File: sample-apps/segmentation_spleen_tta/validate_TTA.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import argparse
```
Comment (SachidanandAlle) [inline sample-apps/segmentation_spleen_tta/validate_TTA.py:1]: lets think about where to put these test utils.. for now let it be part of app folder

File: sample-apps/segmentation_spleen_tta/README.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Segmentation - Spleen
```
Comment (SachidanandAlle) [inline sample-apps/segmentation_spleen_tta/README.md:1]: what is blocking to use the existing segmentation_spleen app for adding tta scoring..? why new APP...?

how about adding this into generic_segmentation app directly?

Comment (diazandr3s) [reply sample-apps/segmentation_spleen_tta/README.md:1]: Thanks for all your comments! I agree we should include all this in the generic App. My plan before Friday is to have a single deepedit generic App once we complete the updates on the deepedit transforms. For now, let's have this as a separate App.

File: monailabel/utils/scoring/test_time_augmentation.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+from monai.transforms.compose import Compose
+from monai.transforms.inverse import InvertibleTransform
+from monai.transforms.inverse_batch_transform import BatchInverseTransform
+from monai.transforms.transform import Randomizable
+from monai.transforms.utils import allow_missing_keys_mode
+from monai.utils.enums import CommonKeys, InverseKeys
+from monai.utils.module import optional_import
+
+if TYPE_CHECKING:
+    from tqdm import tqdm
```
Comment (SachidanandAlle) [inline monailabel/utils/scoring/test_time_augmentation.py:28]: lets not go this route for simple dependencies.. 
you can add tqdm in the requirements.txt if its not installed.. and you don't need optional import

File: monailabel/utils/scoring/test_time_augmentation.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+            mode, mean, std, vvc = tt_aug(test_data)
+    """
+
+    def __init__(
+        self,
+        transform: InvertibleTransform,
+        batch_size: int,
+        num_workers: int,
+        inferrer_fn: Callable,
+        device: Union[str, torch.device] = "cpu",
```
Comment (SachidanandAlle) [inline monailabel/utils/scoring/test_time_augmentation.py:96]: default cpu?

File: sample-apps/segmentation_spleen_tta/lib/activelearning.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+        images = datastore.get_unlabeled_images()
+
+        if not len(images):
+            return None
+
+        tta_scores = {image: datastore.get_image_info(image).get("vvc_tta", 0) for image in images}
+
+        # PICK RANDOM IF THERE IS NOT VVC_TTA SCORES!!
+        if tta_scores[images[0]] == 0:
```
Comment (SachidanandAlle) [inline sample-apps/segmentation_spleen_tta/lib/activelearning.py:59]: better to do `if sum(tta_scores.values()) == 0:`

**Review by SachidanandAlle — APPROVED** [review]

**Pull request merged**